Table Panel: Enable Text Wrapping

codeincarnate opened the pull request:
~~Code isn't really working at this point, but have a workable approach identified I think~~. This renders text content to an offscreen canvas, calculates lines numbers and based on line height calculates total row height. This can then scale the row/cell height appropriately and also give the correct item height to the virtualizer.

Image cell would need a different approach (just image height I guess? 🤷‍♂️ ). Very very much a work in progress.

Fixes #25623 

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

codeincarnate committed:
Calculate row height

codeincarnate committed:
Move things around

codeincarnate committed:
Update getItemSize to use text bounding box

codeincarnate commented:
Progress on this front 🥳 

https://github.com/grafana/grafana/assets/199847/8f2ea97c-4b5c-42a7-80e6-ea0d7f274f01

codeincarnate commented:
Ok, this took a bit too long 😬 but finally getting through some performance testing and with some good news

This approach is using `measureText` in order approximate the size of the text on screen. Text is laid out on one line, the width of the line is measured and then the height is obtained by dividing by the line height to get an approximate line count. 

The downside being that `measureText` isn't necessarily performant, and indeed this does dominate the runtime when using this strategy for variable height (seems like in the neighborhood of 50% to 60% overhead).

![Screenshot 2024-05-07 at 10 47 21 AM](https://github.com/grafana/grafana/assets/199847/9bbf57b6-56c8-4afb-9924-c998a01440c5)


---

However, because we end up drawing less on screen we actually end up mostly balancing out the runtime cost of this, with 7% of time being spent rending with text wrapping enabled and  37% of the time spent rending with text wrapping disabled:

### Text Wrapping Enabled

![Screenshot 2024-05-07 at 10 47 12 AM](https://github.com/grafana/grafana/assets/199847/210208d7-1927-4c99-b896-741e4c3405b6)


### Text Wrapping Disabled

![Screenshot 2024-05-07 at 11 07 17 AM](https://github.com/grafana/grafana/assets/199847/1b67a4d8-7907-4eaa-ab69-a1323126c74a)

In fact this being the case we may actually be able to get **better** interactivity relative to non-wrapped text for this by moving to a web worker and moving work off of the main thread.

So that being said, going to keep going with this as this most certainly seems like a viable strategy for estimating cell height when wrapped.

leeoniya commented:
is this dependent on the number of rows/columns? how many rows/columns is this perf testing?

codeincarnate commented:
This is testing with 5000 rows with 1 to 20 paragraphs of Lorem Ipsum text, currently two columns but I can see if more makes a difference but I suspect it won't change too much (other than strange text wrapping 😆)

codeincarnate committed:
Update types

codeincarnate committed:
Cleanups and reminders

codeincarnate committed:
Calculate line height

codeincarnate committed:
Update line height calculation

codeincarnate committed:
Remove debugging

codeincarnate committed:
Add cell option editing

codeincarnate committed:
Prettier

codeincarnate committed:
Merge branch 'main' into codeincarnate/ftw

codeincarnate committed:
Use field configured for text wrapping

baldm0mma committed:
Merge branch 'main' into codeincarnate/ftw

codeincarnate committed:
Add TODO

leeoniya commented on packages/grafana-ui/src/components/Table/RowsList.tsx:
```diff
@@ -333,3 +377,45 @@ export const RowsList = (props: RowsListProps) => {
+  defaultRowHeight: number
+) {
+  let width = 300;
+  if (typeof headerGroups[0].headers[1].width === 'number') {
```
might be easier to read if `headerGroups[0].headers[1]` was pulled into a `const header = ` above

leeoniya replied:
this whole if/else block can probably be replaced with `Number(headerGroups[0].headers[1])`

codeincarnate replied:
Oops, and this should probably be updated to grab the correct header as well 😆

codeincarnate replied:
OK, simplified this down 😄

codeincarnate committed:
Make sure column configuration is correct

codeincarnate commented:
Here's what is outstanding for this:

- [ ] Add configuration for colored background and color text
- [ ] Fix hover behavior when text wrapping is enabled (open question as to what should be done when one cell has more text, but will _probably_ disable)

codeincarnate committed:
Update height heuristic and hover behavior

codeincarnate committed:
Disable overflow on hover with text wrapping

codeincarnate commented:
I've update the hover behavior and that's solid now. Currently still having heuristic issues. When actually laying out line by line we manage to undercount the actual lines still 😢. Going to brainstorm with @grafana/dataviz-squad tomorrow to get more ideas

codeincarnate committed:
Update heuristic

codeincarnate committed:
Clean things up

codeincarnate committed:
Color background cell options

codeincarnate committed:
Fix tests

codeincarnate committed:
Prettier

codeincarnate committed:
React deps

codeincarnate committed:
Remove old hook dep

codeincarnate committed:
Fix type errors

codeincarnate committed:
Update label and description for editor

codeincarnate committed:
Fix non-wrapped case

codeincarnate committed:
Make sure color background works

codeincarnate commented:
Ok, have color background enabled too. Believe I have the heuristic in a good spot as well. Would be very curious to hear what @leeoniya's (and anyone else's) thoughts are on that 😄

codeincarnate committed:
Prettier

nmarrs committed:
Merge remote-tracking branch 'origin' into codeincarnate/ftw

nmarrs commented on packages/grafana-ui/src/components/Table/RowsList.tsx:
```diff
@@ -68,13 +71,28 @@ export const RowsList = (props: RowsListProps) => {
   const theme = useTheme2();
   const panelContext = usePanelContext();
 
+  // Create off-screen canbas for measuring rows for virtualized rendering
```
```suggestion
  // Create off-screen canvas for measuring rows for virtualized rendering
```

nmarrs commented on packages/grafana-ui/src/components/Table/RowsList.tsx:
```diff
@@ -218,6 +237,18 @@ export const RowsList = (props: RowsListProps) => {
+      fieldOptions.cellOptions !== undefined &&
+      (fieldOptions.cellOptions.type === TableCellDisplayMode.Auto ||
+        fieldOptions.cellOptions.type === TableCellDisplayMode.ColorBackground) &&
+      fieldOptions.cellOptions.wrapText
```
it might be nice to break this out into a few separate consts defining what these checks are 😬

codeincarnate replied:
Done!

nmarrs commented on packages/grafana-ui/src/components/Table/RowsList.tsx:
```diff
@@ -218,6 +237,18 @@ export const RowsList = (props: RowsListProps) => {
       };
     }
+
+    // TODO: Also add option for colored background cell,
```
do we have a follow-up github issue capturing this?

codeincarnate replied:
Actually added this 😄, probably want to also hit colored text too but I can make an issue for that

nmarrs commented on packages/grafana-ui/src/components/Table/RowsList.tsx:
```diff
@@ -68,13 +71,28 @@ export const RowsList = (props: RowsListProps) => {
+  // nor is it a part of the jest-canvas-mock package
+  let osContext = null;
+  if (window.OffscreenCanvas !== undefined) {
+    osContext = new OffscreenCanvas(256, 1024).getContext('2d');
```
should we define these numbers as consts / what they represent? 🤔

nmarrs commented on packages/grafana-ui/src/components/Table/RowsList.tsx:
```diff
@@ -333,3 +399,52 @@ export const RowsList = (props: RowsListProps) => {
+    const lines = [];
+    let currentLine = '';
+
+    // Let's just wrap the liens and see how well the measurement works
```
```suggestion
    // Let's just wrap the lines and see how well the measurement works
```

nmarrs commented on packages/grafana-ui/src/components/Table/RowsList.tsx:
```diff
@@ -333,3 +399,52 @@ export const RowsList = (props: RowsListProps) => {
 };
+
+/**
+ * Calculate an esimated bounding box for a block
```
```suggestion
 * Calculate an estimated bounding box for a block
```

nmarrs commented on packages/grafana-ui/src/components/Table/RowsList.tsx:
```diff
@@ -25,6 +25,8 @@ import { TableStyles } from './styles';
 import { calculateAroundPointThreshold, getCellColors, isPointTimeValAroundTableTimeVal } from './utils';
 
+const LINE_SCALE_FACTOR = 1.12;
+const LOW_LINE_PAD = 32;
```
I think we could move these to be defined as part of the getTextBoundingBox function as that's the only place they're being referenced

nmarrs reviewed (commented):
Overall this is looking solid Kyle! I'm having a bit of trouble though locally to see the "auto" cell height option come up so i've only been able to test toggling on the "wrap text" option so far 🤔 
<img width="1472" alt="Screenshot 2024-05-13 at 8 33 25 PM" src="https://github.com/grafana/grafana/assets/22381771/93b52b9f-c55f-4ed8-a585-e50c72a9e39e">


[This gdev dashboard (Table panel - Pagination)](http://localhost:3000/d/1YGJxPUnz/table-panel-pagination?orgId=1) is a good way to test this out

codeincarnate committed:
Address review comments

codeincarnate commented:
@nmarrs I addressed your review comments. So effectively the wrap text option on the cell overrides the height property 😄. Granted the presents a somewhat confusing user experience. It was done this way to keep the same pattern that was used for row coloring by using a field override. 

In general it makes in straightforward to define which field should be used for sizing, but not sure if the experience is too confusing? It's something we could potentially rework but I'm definitely open to different options 😄

nmarrs committed:
Merge remote-tracking branch 'origin' into codeincarnate/ftw

nmarrs committed:
fix prettier

nmarrs reviewed (commented):
Hey @codeincarnate - thanks for updating / addressing code fixes! I think I understand better what is happening and I think things are working as expected for the most part.

I think the question that I have is which field is selected to drive the cell height, IMO it probably should be the one that has the longest text? In the example from [this panel from the gdev dashboard](https://github.com/grafana/grafana/files/15315693/debug-No.pagination.footer-2024-05-14.19_52_26.json.txt) the panel it using `property name` vs `description` which is way longer 🤔 

I think you mentioned the selection of the field used to drive the overall table cell height can be driven by field overrides? We may want to have some type of inline documentation indicating this and / or pick the field with the most amount of text as the default. What do you think?


https://github.com/grafana/grafana/assets/22381771/c2510949-86c7-423b-85fc-b703f7a26199

codeincarnate commented:
> pick the field with the most amount of text as the default

@nmarrs I think that's a great suggestion! It won't be perfect but we can use a sampling of rows as a heuristic. I think it's quite likely that most of the time the longest text field will be the one that's desirable to wrap

codeincarnate committed:
Add heuristic for field sizing

codeincarnate committed:
Merge branch 'codeincarnate/ftw' of github.com:grafana/grafana into codeincarnate/ftw

codeincarnate committed:
Fix up logic

codeincarnate committed:
Prettier

codeincarnate committed:
Fix test

codeincarnate committed:
Oh prettier 🙈

codeincarnate commented:
@nmarrs I implemented your suggestion if you'd like to test it out 😄

nmarrs committed:
Merge remote-tracking branch 'origin' into codeincarnate/ftw

nmarrs reviewed (approved):
nice! I really like the change where we have the default text wrap field driven by the longest field :) 

There are still some edge cases where this isn't working perfectly, perhaps adding a bit of additional margin (1-2 line space in addition to what is calculated) wouldn't hurt as currently the text mostly feels crowded / a few places where calcs did not account for 1-2 lines being cutoff in top / bottom - I'm good with this being handled in a follow-up issue though if desired

[see this debug panel dashboard for reference.txt](https://github.com/grafana/grafana/files/15341183/debug-Reproduced.with.embedded.data-2024-05-16.14_17_35.json.txt)



https://github.com/grafana/grafana/assets/22381771/052ba66c-1627-44f2-ad72-87966c75ca62



Great job on this! 🍬🏅

torkelo commented:
I have always thought the way to support text wrapping is to have an option that 
* skips the virtualized list, uses normal table DOM elements
* enforce pagination 
* disable support for chart & bar gauge cells 

The lack of support for cells that need dimensions is tricky, maybe a new visualization (Wrapping table)

torkelo reviewed (changes requested):
Crashed on numeric fields 
![Screenshot 2024-05-17 at 08 26 16](https://github.com/grafana/grafana/assets/10999/4e638a8f-6ca9-4d85-a532-fda054c6ced0)

Missing wrap text option for "Colored text" cell type. 

I do not think the text wrapping is working great, many cells lack padding 

![Screenshot 2024-05-17 at 08 53 32](https://github.com/grafana/grafana/assets/10999/7a732194-01bb-4134-9481-9055fa37b7d6)

codeincarnate commented:
@torkelo I'll take a look at the crash and tweaking the wrapping 😄 

In terms of the other approach I think it would certainly be worth a look though that seems like a heavier lift overall than this approach. I believe logs is taking a similar approach to this so seems viable until we can look at doing a bigger re-work.

codeincarnate committed:
Don't wrap text on non-string fields

codeincarnate committed:
Merge branch 'codeincarnate/ftw' of github.com:grafana/grafana into codeincarnate/ftw

codeincarnate committed:
Add wrapping to color text cell

codeincarnate committed:
Prettier

codeincarnate commented:
@torkelo I've updated to fix the bug, add support for color text cells, and the text wrapping. Happy to make more updates in that regard. Also, if you could share your data set I would be happy to do further testing there 😄

codeincarnate commented:
Hey @torkelo, wanted to see if you had any further feedback on the most recent changes. If you could take a look it would be greatly appreciated. Thanks! 🙏

torkelo commented:
@codeincarnate sorry, partly out if office this week so been struggling finding time testing this 

But if you and Nathan feel good about this let's merge it 👍

torkelo commented:
@codeincarnate took it for another test spin, still some big issues here for me. 

I cannot figure out / see a way to enable text wrapping without also enabling colored text or colored background (could not find it in the default auto cell type, nor as a override option). 

Having only one field control the height makes it very inconsistent when you have two fields or more that sometimes have longer or shorter content. 
<img width="896" alt="Screenshot 2024-05-23 at 11 05 38" src="https://github.com/grafana/grafana/assets/10999/487d0118-0896-4914-b1b0-a0179d8b455b">

codeincarnate committed:
Fix option not showing for auto cell type

codeincarnate commented:
@torkelo no worries at all! Thanks for taking another look at this 🙏 

In terms of the auto cell there was an accident in the logic to display the editor for that so my apologies there and should be good now 😄 

As for the height, that's definitely a fair criticism of this approach. I went with this for performance considerations because we're doing a simulated wrapping of text in the background to estimate height for the virtualizer. Selecting one column makes it faster to get an estimated height for the virtualizer. 

We could potentially enable for the table, we would have to look at every string column to get the maximum height for that row. I'm certainly not opposed to the idea but I think we'd start hitting some performance issues.

At the moment we're getting a decent trade-off in saved rendering relative to estimating the height. Not sure how that approach would pan out if we scan all columns. It's unfortunate in cases like above where two columns have similar text lengths. I'm not sure if we can estimate all columns while maintaining good performance in the context of our current virtualizer. Very interested on to hear thoughts on this!

torkelo commented on packages/grafana-ui/src/components/Table/RowsList.tsx:
```diff
@@ -333,3 +402,54 @@ export const RowsList = (props: RowsListProps) => {
+ * Calculate an estimated bounding box for a block
+ * of text using an offscreen canvas.
+ *
+ * TODO: Move this 🚚
```
todo, I assume before merge :)

codeincarnate replied:
Thanks! 🙏  yes :)

torkelo commented on packages/grafana-ui/src/components/Table/Table.tsx:
```diff
@@ -287,6 +288,56 @@ export const Table = memo((props: Props) => {
+    const stringFields = data.fields.filter((field: Field) => field.type === FieldType.string);
+
+    if (stringFields.length >= 1) {
+      const numValues = stringFields[0].values.length;
+      let longestLength = 0;
+
+      // If we have less than 20 values we assume
+      // that the first record is representative
+      // of the overall data
+      if (numValues <= 30) {
+        for (const field of stringFields) {
+          const fieldLength = field.values[0].length;
+          if (fieldLength > longestLength) {
+            longestLength = fieldLength;
+            longestField = field;
+          }
+        }
+      }
+      // Otherwise we randomly sample 3 values and take
+      // the mean length
+      else {
+        for (const field of stringFields) {
+          // This could result in duplicate values but
+          // that should be fairly. This could potentially
+          // be improved using a Set datastructure but
+          // going to leave that one as an exercise for
+          // the reader to contemplate and possibly code
+          const val1 = sample(field.values);
+          const val2 = sample(field.values);
+          const val3 = sample(field.values);
+          const meanLength = (val1.length + val2.length + val3.length) / 3;
+
+          if (meanLength > longestLength) {
+            longestLength = meanLength;
+            longestField = field;
+          }
+        }
+      }
+    }
+  }
```
This feels like it does not belong in the main render method and seems like a function that should have 1.-2 tests

torkelo commented on packages/grafana-ui/src/components/Table/Table.tsx:
```diff
@@ -287,6 +288,56 @@ export const Table = memo((props: Props) => {
+          // be improved using a Set datastructure but
+          // going to leave that one as an exercise for
+          // the reader to contemplate and possibly code
+          const val1 = sample(field.values);
+          const val2 = sample(field.values);
+          const val3 = sample(field.values);
```
samples = sampleSize(field.values, 3)

[sample](https://lodash.com/docs/4.17.15#sampleSize)

torkelo reviewed (commented):
I am torn here, this approach really depends on how common it will be for a single field to require wrapping as it does not work if two or more fields require wrapping.

Have we asked in https://github.com/grafana/grafana/issues/25623 (or customers) what scenario they are in need of text wrapping (single field vs multi field)

appi101 commented:
> I am torn here, this approach really depends on how common it will be for a single field to require wrapping as it does not work if two or more fields require wrapping.
> 
> Have we asked in https://github.com/grafana/grafana/issues/25623 (or customers) what scenario they are in need of text wrapping (single field vs multi field)  

Fwiw, the single field scenario is what would be most beneficial. My use case has partially structured data and then one free form text field.

codeincarnate committed:
Move longest field guessing into function

codeincarnate committed:
Merge branch 'main' into codeincarnate/ftw

codeincarnate committed:
Clean things up

codeincarnate commented:
@torkelo definitely understand. One of the guiding use cases for this was logging (and similar scenarios) where there's typically one field much longer than the rest. I'm not sure about how prevalent the scenario of multiple text fields of varying size is. I've reached out in #25623 to gather feedback in that regard

From what I remember from the customer scenarios I've looked at is that the case I outlined above was pretty common. Can't link those here naturally, but I'll do my due diligence on that one and see if I can square that away 😄 

Otherwise, in the mean time I've refactored things based on your comments. Going to add a few tests for the function above as well.

codeincarnate committed:
Add tests

codeincarnate committed:
Make sure text won't flake

codeincarnate committed:
Prettier

codeincarnate committed:
Remove spurious import

codeincarnate committed:
Ignore any in this case

torkelo commented:
@codeincarnate I see, well we can always label it as experimental to and merge it to gather feedback :)

codeincarnate committed:
Add alpha label

codeincarnate committed:
Prettier

codeincarnate committed:
Fix typecheck

codeincarnate commented:
@torkelo I've added an alpha label to the option for wrapping. Would very much appreciate another look 🙏 

Also looking into options we could follow up with to text-wrap for multi-column scenarios. I think that would be a good follow up to this and seems reasonable at the moment to collect what feedback we can regarding this feature 😄

codeincarnate committed:
Fix crash when sampling when there are undefined records

codeincarnate committed:
Merge branch 'main' into codeincarnate/ftw

codeincarnate committed:
Update heuristic to take into account long strings

codeincarnate committed:
Prettier

codeincarnate committed:
Update scale factors

codeincarnate committed:
Update field index selection

codeincarnate committed:
Prettier
